### PR TITLE
Android life cycle improvements and Gradle integration

### LINKED ISSUE #14568: Runtime problems on Android
Now that the [build works](https://github.com/servo/servo/issues/13154), you can run Servo on Android again after applying these patches:

- #14528 ✓ 
- #14530 ✓ 
- servo/webrender#626 ✓ 
- servo/webrender#627 ✓ 
- servo/webrender#632 ✓ 
- servo/webrender#653 ✓ 
- servo/webrender#654 ✓ 

When you start Servo on Android, it produces a gray screen. This is because a [missing `WindowProxy` implementation](https://github.com/servo/servo/blob/8b69e73594647319e95bd0fd36c2addabcee1e5d/ports/glutin/window.rs#L760) on Android.

On any input event, the gray screen disappears, and Servo itself starts. At the moment, however, this results in a mostly white screen, with one or two elements visible occasionally. There are no errors, the main loop is running, and I don't really see anything particularly interesting in the logs either. Also some sites may throw shader compiler errors too.

Here's a screenshot of `https://www.google.com`, there, we only have a scrollbar, the menu button and the login button's text:

screen

### PR DESCRIPTION
**Pull request by MortimerGoro**
This PR includes Android life cycle Improvements and Gradle integration for android packaging.

Android life cycle improvements are implemented in both the new [MainActivity](https://github.com/servo/servo/compare/master...MortimerGoro:android_improvements?expand=1#diff-f43708b102e98272c2af7454b8846927) and native code in this related PR: https://github.com/servo/glutin/pull/117 
- Properly handle the life cycle of the Android Activity: manage EGLContext lost & restore, and animation loop pause/resume when the app goes to background/foreground or orientation changes. In the current upstream Servo crashes when the app goes to background, activity stops or changes orientation
- Handle event loop awake for Servo Animation loop
- Handle screen resize & orientation events
- Implement new keep_screen_on preference which keeps Android device's screen from ever going to sleep: very useful for games or WebVR
- Implement a full screen mode enabled by preference: very useful for games or WebVR
- Implement new shell.native-orientation preference which allows to lock the Activity to a specific orientation
- Automatically sync new android assets to sdcard when the Android version code is changed.  In the current upstream only the existence of the folder is check but resources are not updated

ofscreen_gl_context is updated to fix this: https://github.com/emilio/rust-offscreen-rendering-context/pull/83

This PR integrates Gradle build system for android packaging. Most of the code is implemented in this [build.gradle](https://github.com/servo/servo/compare/master...MortimerGoro:android_improvements?expand=1#diff-89cdb9324addb994cdba0a158b209547) . We can get rid of [build-apk](https://github.com/servo/servo/compare/master...MortimerGoro:android_improvements?expand=1#diff-40f5a7cf22f94aad059b2c1795347f5e) and manual jar dependency copying in the [package_commands.py](https://github.com/servo/servo/compare/master...MortimerGoro:android_improvements?expand=1#diff-0d425b142c8d10ae6ac1f3711fb5c23a). The correct version of gradle is automatically downloaded using the gradlew wrapper.

Some improvements:
- Allows to include more complex android dependencies/SDKs like AARs, manifest auto-merging and more.
- Improved packaging process: The gradle project is always in the same folder, it uses relative paths for everything (assets, native libraries) and outputs the apk into the correct target folder in servo. In the current upstream, ant/python build system copies the manifest, project, resources and jars each time so you end with multiple copies of the same files.
- Improved dependency declaration. We do not have to manually copy jar dependencies in the python script anymore. The gradle build scripts itself is able to search for the dependencies in the correct servo target folder.
- Supports packaging apks with different architectures: armeabi, armeabi-v7a, aarch64. We still need to fix some native servo compilation issues with armeabi-v7a, aarch64 due to dependencies which use `make`. I'll push this changes in a separate PR of the python build files but the gradle file is already ready to handle that.
- We can easily create product flavors for different versions of Servo. For example a default browser, a WebVR browser with additional dependencies or a Servo android Webview component
- Fixes minor.major.52 build error when blurdroid cargo dependency is compiled using java8 (the default in new Linux machines). The gradle build file enables the new Jack compiler which supports Java8 dependencies.
- The project can be opened with Android Studio and run the brand new GPU debugger on any Android phone. I'll add some docs in the Wiki about this.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #14568 (github issue number if applicable).

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15773)

### CODE REVIEW

**Comment by larsbergstrom** [comment]
Awesome work, @MortimerGoro!

@fabricedesre Would you mind doing the initial review of this? :-)

**Comment by fabricedesre** [comment]
@larsbergstrom sure, will do.

File: support/android/apk/app/build.gradle
Comment (fabricedesre) [inline support/android/apk/app/build.gradle]: can't we do something like `for (debug in [true, false])` ?

File: support/android/apk/app/src/main/AndroidManifest.xml
Comment (fabricedesre) [inline support/android/apk/app/src/main/AndroidManifest.xml]: Use `auto` as a value for installLocation, and move it to its own line.

File: support/android/apk/app/src/main/java/com/mozilla/servo/MainActivity.java
Comment (fabricedesre) [inline support/android/apk/app/src/main/java/com/mozilla/servo/MainActivity.java]: this can be simplified once we only target sdk >= 21

File: support/android/apk/app/src/main/java/com/mozilla/servo/MainActivity.java
Comment (fabricedesre) [inline support/android/apk/app/src/main/java/com/mozilla/servo/MainActivity.java]: Same here.

Also the style in this file is not consistent. Either always put the opening brace on its own line or on the same line as the method declaration, but don't mix and match.

File: support/android/apk/app/src/main/java/com/mozilla/servo/MainActivity.java
Comment (fabricedesre) [inline support/android/apk/app/src/main/java/com/mozilla/servo/MainActivity.java]: let's remove that.

File: support/android/apk/app/src/main/java/com/mozilla/servo/MainActivity.java
Comment (fabricedesre) [inline support/android/apk/app/src/main/java/com/mozilla/servo/MainActivity.java]: nit: like for setFullScreen, you could get the view inside this method instead of passing it as a parameter.

File: support/android/apk/app/src/main/java/com/mozilla/servo/MainActivity.java
Comment (fabricedesre) [inline support/android/apk/app/src/main/java/com/mozilla/servo/MainActivity.java]: Please use the constants from https://developer.android.com/reference/android/view/View.OnSystemUiVisibilityChangeListener.html

File: support/android/apk/app/src/main/java/com/mozilla/servo/MainActivity.java
Comment (fabricedesre) [inline support/android/apk/app/src/main/java/com/mozilla/servo/MainActivity.java]: nit: `reader = new BufferedReadre(new InputStreamReader(is));` so you won't have to deal manually with isr.

File: support/android/apk/app/src/main/java/com/mozilla/servo/MainActivity.java
Comment (fabricedesre) [inline support/android/apk/app/src/main/java/com/mozilla/servo/MainActivity.java]: Please use {} even for single line blocks.

File: support/android/apk/app/src/main/java/com/mozilla/servo/MainActivity.java
Comment (fabricedesre) [inline support/android/apk/app/src/main/java/com/mozilla/servo/MainActivity.java]: Maybe returning an empty JSONObject would be better?

File: support/android/apk/gradle.properties
Comment (fabricedesre) [inline support/android/apk/gradle.properties]: nit: remove.

File: support/android/apk/gradle/wrapper/gradle-wrapper.properties
Comment (fabricedesre) [inline support/android/apk/gradle/wrapper/gradle-wrapper.properties]: nit: not really useful ;)

**Review by fabricedesre — CHANGES_REQUESTED** [review]
Thanks for doing that, this is super nice!
I left mostly nits, except the change to a minimum sdk version that will simplify the code.

File: support/android/apk/app/build.gradle
Comment (MortimerGoro) [inline support/android/apk/app/build.gradle]: I agree that setting target 21 would simplify parts of the code (is the minimum api level for native arm64 builds too). But Jelly bean + KitKat still has around 35-40% market share (https://developer.android.com/about/dashboards/index.html). What do you think about this @larsbergstrom ?

**Comment by MortimerGoro** [comment]
Thanks for the review @fabricedesre 

I pushed all the requested changed. I'm waiting for https://github.com/servo/glutin/pull/117 to be merged to update the glutin version.

**Review by fabricedesre — APPROVED** [review]

**Comment by fabricedesre** [comment]
hm, I'm not sure why the ccache travis build is failing.

**Comment by MortimerGoro** [comment]
It seems like a random crash. Anyway we need to merge https://github.com/servo/glutin/pull/117 or travis build will fail too in another step.

**Comment by fabricedesre** [comment]
Not sure it's random, since I re triggered the build and it failed again. But yep, let's wait for the glutin changes first!

**Comment by MortimerGoro** [comment]
@larsbergstrom Thanks for merging glutin PR https://github.com/servo/glutin/pull/117 But no version has been uploaded to crates.io (I thought that it was automatic). Can you publish the new version? I can switch to a github based dependency if required.

@fabricedesre I have updated the PR with some java/rust changes related to android resources/data paths. Upgrading the target SDK breaks permissions for writing in /sdcard/. I have switched to the recommended getExternalFilesDir paths.

**Comment by fabricedesre** [comment]
@MortimerGoro thanks for the update! Can you *not* squash when doing that though, it's easier to look at interdiffs when reviewing incremental changes.

**Comment by mbrubeck** [comment]
Published https://crates.io/crates/servo-glutin/0.9.1

**Comment by MortimerGoro** [comment]
@mbrubeck Thanks!

@fabricedesre Ok, the correct way is to squash the commits after the PR is approved. Sorry for the inconveniences. Let's do that way ;)

**Comment by mmatyas** [comment]
Finally had some time to try this out, really nice work!

File: support/android/apk/app/build.gradle
Comment (mmatyas) [inline support/android/apk/app/build.gradle]: Older devices are still widely used (think developing countries and regional differences), and several of our test devices also run Android 4.4. I'd like API 18 support too.

**Comment by MortimerGoro** [comment]
Thanks @mmatyas 

> Older devices are still widely used (think developing countries and regional differences), and several of our test devices also run Android 4.4. I'd like API 18 support too.

I've upgraded the PR to support API 18. The android version compatibility code is much simpler now as requested by @fabricedesre. Only a full screen View flag was breaking compatibility with API 18, easy to handle.

**Comment by MortimerGoro** [comment]
Not sure why the Travis CI build is failing:
Extracting Cargo nightly...
```
Cargo ready.
cargo build -v
    Updating registry `https://github.com/rust-lang/crates.io-index`
    Updating git repository `https://github.com/servo/webrender`
    Updating git repository `https://github.com/mmatyas/android-rs-injected-glue`
    Updating git repository `https://github.com/pcwalton/signpost.git`
    Updating git repository `https://github.com/energymon/energymon-rust.git`
    Updating git repository `https://github.com/servo/gaol`
    Updating git repository `https://github.com/servo/rust-azure`
    Updating git repository `https://github.com/servo/rust-mozjs`
    Updating git repository `https://github.com/servo/angle`
    Updating git repository `https://github.com/servo/osmesa-src`
    Updating git repository `https://github.com/browserhtml/browserhtml`
    Updating git repository `https://github.com/energymon/energymon-sys.git`
    Updating git repository `https://github.com/servo/mozjs`
error: process didn't exit successfully: `rustc -vV` (exit code: 127)
--- stderr
rustc: error while loading shared libraries: librustc_driver-b65378ba5ffb99d4.so: cannot open shared object file: No such file or directory
```
Similar error info tracked here https://github.com/rust-lang/rust/issues/24677, but this PR shouldn't affect to that. Any ideas?

**Comment by fabricedesre** [comment]
@larsbergstrom any idea about the CI failure?

**Comment by larsbergstrom** [comment]
Hrm, I haven't seen that before. I wonder if it's running out of disk space?

**Comment by larsbergstrom** [comment]
I triggered a re-run and it's now passing. @fabricedesre Is this ready to land or does it need a squash?

**Comment by fabricedesre** [comment]
@larsbergstrom that would be nice to squash before merging

**Comment by MortimerGoro** [comment]
Probably https://github.com/servo/servo/pull/15911 has helped to fix the issue

@fabricedesre I did the squash and added r=fabricedesre

**Comment by larsbergstrom** [comment]
@bors-servo retry
- git lol

**Comment by larsbergstrom** [comment]
```

FAILURE: Build failed with an exception.

* Where:
Build file '/home/servo/buildbot/slave/android/build/support/android/apk/app/build.gradle' line: 1

* What went wrong:
A problem occurred evaluating project ':app'.
> java.lang.UnsupportedClassVersionError: com/android/build/gradle/AppPlugin : Unsupported major.minor version 52.0

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
```

**Comment by MortimerGoro** [comment]
@larsbergstrom That's because the build machine needs: sudo apt-get install openjdk-8-jdk. Where is the right place to add that?

**Comment by larsbergstrom** [comment]
@MortimerGoro https://github.com/servo/saltfs/blob/master/servo-build-dependencies/android.sls#L26

**Comment by larsbergstrom** [comment]
@bors-servo retry
- https://github.com/servo/saltfs/pull/617 has landed and been deployed

**Comment by fabricedesre** [comment]
Hm, still failing with the same error: http://build.servo.org/builders/android/builds/5747/steps/compile__1/logs/stdio

**Comment by larsbergstrom** [comment]
Yeah, I'm curious about this - did the highstate from the server and it thinks the client is up to date, but on the machine itself:
```
root@IP:~# dpkg -l | grep jdk
ii  default-jdk                          2:1.7-51                            amd64        Standard Java or Java compatible Development Kit
ii  openjdk-7-jdk:amd64                  7u121-2.6.8-1ubuntu0.14.04.1        amd64        OpenJDK Development Kit (JDK)
ii  openjdk-7-jre:amd64                  7u121-2.6.8-1ubuntu0.14.04.1        amd64        OpenJDK Java runtime, using Hotspot JIT
ii  openjdk-7-jre-headless:amd64         7u121-2.6.8-1ubuntu0.14.04.1        amd64        OpenJDK Java runtime, using Hotspot JIT (headless)
```

**Comment by MortimerGoro** [comment]
@fabricedesre @larsbergstrom This is also required: https://github.com/servo/saltfs/pull/618

**Comment by mmatyas** [comment]
When will this land?

**Comment by MortimerGoro** [comment]
@mmatyas We are waiting for this: https://github.com/servo/saltfs/issues/462#issuecomment-287412943

**Comment by aneeshusa** [comment]
https://github.com/servo/saltfs/pull/629 should fix installing OpenJDK 8 on 14.04.

**Comment by MortimerGoro** [comment]
@aneeshusa Thanks!

@fabricedesre @larsbergstrom Rebase ready. This should be ready for test & merge.

**Comment by larsbergstrom** [comment]
Deployed the update to the machines!

@bors-servo r=larsbergstrom,fabricedesre

**Comment by aneeshusa** [comment]
Java 8 is deployed, but I checked the builders and Java 7 is still the default. We'll need to use `update-alternatives` to set Java 8 as the default, I'm guessing - I'll look into it after lab.

**Comment by larsbergstrom** [comment]
@Manishearth Did you say that you ended up redeploying the update to 8 from scratch on the machines? Or did you have to roll it back? I didn't see any issues or mentions, but may have missed it and don't want to break something new.

**Comment by aneeshusa** [comment]
servo/saltfs#618 got rolled back as servo/saltfs#634, which has been deployed to unblock other PRs. It seems like the new version of the Android tools has an incompatible directory layout, so updating the build tools for this PR will require some additional Salt code beyond just bumping the version. Worst case, I have some ideas about how to have both versions of the Android tools installed simultaneously.

**Comment by Manishearth** [comment]
I reverted this change and deployed the reverted version.

**Comment by aneeshusa** [comment]
To be clear, the installation of OpenJDK 8 was not reverted and is still deployed on the cross builders, but Java 7 is still the default Java. We either need another saltfs update to set Java 8 as the default, or use environment variables in this PR to use Java 8 (e.g. maybe via `JAVA_HOME`).

**Comment by larsbergstrom** [comment]
@Manishearth @aneeshusa So is the next step here to reland servo/saltfs#618  but instead of doing an incremental deploy just burn down the cross builders and re-create them from scratch to avoid two versions of Java? b/c it seems like that works fine in the vagrant setup where they're provisioned from scratch, so I assume this is mainly a problem with incremental upgrades.

cc @edunham

**Comment by aneeshusa** [comment]
@larsbergstrom I don't think that's quite right, there are two separate upgrades at play here:

1. OpenJDK 8: The existing builders already had JDK 7 installed, and servo/saltfs#617 and servo/saltfs#629 added OpenJDK 8 to our cross builders. JDK 7 is still the default, which I believe is causing JDK 8 to not be found for this PR. Options I see here are a) set the default JDK to Java 8 (I'm not sure if this will break our existing Android build) and possibly remove JDK 7 and b) use environment variables (`JAVA_HOME` I believe) to point this build at Java 8 while letting other builds continue to use Java 7.

2. Android tools: servo/saltfs#618 upgraded our version of the Android tools, which seems to have deployed just fine. However, it looks like the new tools have a different directory layout, and so tools aren't in the same place causing builds to fail. It's also possible the new tools are incompatible with our existing Android build - I'm not sure about this. Options I see here are a) reland 618 but with additional Salt code to fix up the directory paths and hope the new Android tools are compatible with our existing servo Android build, and b) use Salt to install both versions of the tools simultaneously with the old tools as the default, and override the relevant environment variables in this PR to point at the new tools.

**Comment by MortimerGoro** [comment]
@aneeshusa @larsbergstrom BTW:

The ant build system works with Java8 on my Ubuntu 16.04 machine, but java7 must be totally removed or update-aternative called with many commands (It happened to me that ant used Java7 while blurdroid.jar javac compilation used Java8).

About Android sdk tools I have both android-18 and android-25 installed, and I can package with the old and with the new build systems. 

BTW:  is it possible to merge this PR quickly after deploying new Android tools  https://github.com/servo/saltfs/pull/618 and setting Java8 as default (or removing Java7)?. It seems easier than the potential work required to handle both old & new at the same time. The other blocked PR time can be minimal if we can do all the steps fast.

**Comment by aneeshusa** [comment]
@MortimerGoro I've been wanting to extend `./mach bootstrap` with an `--android` flag to set up dependencies for building Servo for Android. If possible, I'd like bootstrapping to not cause invasive changes to a contributor's system that could break their existing setup, e.g. removing Java 7 or changing the default Java. Do you know if having both Java 7 and Java 8 installed and using `JAVA_HOME` (or some other environment variable) to switch between them works with ant and gradle? If not, I think removing Java 7 will be easier to implement on our builders.

**Comment by MortimerGoro** [comment]
@aneeshusa JAVE_HOME works with gradle. Not sure about ant, but it should too. Where are you thinking to add the environment variable? in python scripts or is there a better place to set it?

**Comment by aneeshusa** [comment]
@MortimerGoro I think it makes most sense to set `JAVA_HOME` in saltfs and let `mach` just use whatever is configured at the system level. (In the future, the bootstrapper could provide a hint to mach about which Java to use.) If you confirm that `JAVA_HOME` works for our current builds with ant, I can make a PR to saltfs to set the environment variable.

**Comment by MortimerGoro** [comment]
@aneeshusa I ran some JAVA_HOME tests, more info in https://github.com/servo/saltfs/pull/638

**Comment by aneeshusa** [comment]
FYI, I've started working on a PR to install both sets of Android build tools concurrently - it looks like the only major change is that they removed a directory layer in the archive.

**Comment by aneeshusa** [comment]
WIP PR is up at https://github.com/servo/saltfs/pull/639, I'll keep working on it tomorrow. I haven't tested it at all yet.

**Comment by aneeshusa** [comment]
Once servo/saltfs#639 lands, we'll need to add code to this PR to override the `ANDROID_SDK` (and maybe `PATH`? I'm not sure about this) environment variables to point at the newer SDK version (e.g. override them from [what Buildbot sets](https://github.com/servo/saltfs/blob/de820a3889e2e770281d7adfb46c021e4dcbec0d/buildbot/master/files/config/environments.py#L79)).

@MortimerGoro I think the best place is to override them is in `etc/ci/buildbot_steps.yml`; once this PR is merged, we can update saltfs to make the new SDK version the default and then remove the env var overrides from servo/servo.

Also, please remove `r=fabricedesre` from the commit message, Homu will add it for us automatically.

**Comment by MortimerGoro** [comment]
@aneeshusa I added the ANDROID_SDK override code. Can you check if it's correct? https://github.com/MortimerGoro/servo/blob/5a4a12c20e18dff0dae403e9d85b55341288d7f8/etc/ci/buildbot_steps.yml#L97

PATH variable shouldn't affect.
 
> Also, please remove r=fabricedesre from the commit message, Homu will add it for us automatically.

done

**Comment by aneeshusa** [comment]
I don't believe we do tilde expansion for the home directory, so it's probably safer to use the full `/home/servo/...` form. Otherwise lgtm.

**Comment by MortimerGoro** [comment]
@aneeshusa ok, changed to use full path.

**Comment by aneeshusa** [comment]
servo/saltfs#639 is deployed.

@bors-servo r=larsbergstrom,fabricedesre

**Pull request merged**
